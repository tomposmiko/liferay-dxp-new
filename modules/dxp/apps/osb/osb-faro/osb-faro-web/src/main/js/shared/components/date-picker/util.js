import moment from 'moment';
import {conformsTo, isNil} from 'lodash';

/**
 * Represents a time between two dates.
 * @typedef Range
 * @property {Moment} start - The start of the range.
 * @property {Moment} end - The end of the range.
 */

export function isAboveMaxRange(date, maxRange) {
	return (
		isRange(date) &&
		date.start &&
		date.end &&
		maxRange &&
		moment
			.duration({
				from: date.start,
				to: date.end
			})
			.asDays() > maxRange
	);
}

/**
 * A validation function that determines if a value is a valid
 * moment instance, allowing for null.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
function isMomentOrNil(value) {
	return isNil(value) || moment.isMoment(value);
}

/**
 * A validation function that determines if the value is a valid
 * moment or range.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isDateOrRange(value) {
	return isNil(value) || moment.isMoment(value) || isRange(value);
}

/**
 * Determines if a date is in a {@link Range}.
 * @param {Range|Moment} dateOrRange - The range to check in.
 * @param {Moment} date - The date to check for.
 * @returns {boolean}
 */
export function isInRange(dateOrRange, date) {
	if (moment.isMoment(dateOrRange)) {
		return false;
	}

	const {end, start} = dateOrRange;

	return start && date.isAfter(start) && end && date.isBefore(end);
}

/**
 * A validation function that determines if a value is a valid {@link Range}.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isRange(value) {
	return conformsTo(value, {
		end: isMomentOrNil,
		start: isMomentOrNil
	});
}

/**
 * Adds a date to an existing {@link Range}.
 *
 * This function will try to be smart and first update the
 * start or end if either are null. If both exist, then it will return
 * a new range with the start set to the new date, and the end reset
 * to null.
 * @param {Range} range - The range to update.
 * @param {Moment} date - The date to add to the range.
 * @returns {Range}
 */
export function updateRange(range, date) {
	if (!range.start) {
		return {
			...range,
			start: date
		};
	}

	if (range.start && !range.end) {
		if (date.isSameOrAfter(range.start)) {
			return {
				...range,
				end: date
			};
		} else {
			return {
				end: range.start,
				start: date
			};
		}
	}

	if (date.isSame(range.start) && date.isSame(range.end)) {
		return range;
	}

	return {
		end: null,
		start: date
	};
}
