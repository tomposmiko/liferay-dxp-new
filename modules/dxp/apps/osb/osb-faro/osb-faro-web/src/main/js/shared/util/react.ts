import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
