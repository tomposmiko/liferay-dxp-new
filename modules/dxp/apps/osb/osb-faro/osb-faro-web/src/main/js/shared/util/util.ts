import {Align} from 'metal-position';
import {
	ALIGNMENTS_MAP,
	POSITIONS,
	RangeKeyTimeRanges
} from 'shared/util/constants';
import {flow, get, isFinite, isNil, isString, toLower, trim} from 'lodash';
import {
	RangeSelectors,
	RawRangeSelectors,
	SafeRangeSelectors
} from 'shared/types';

/**
 * Check if the value is blank.
 * @param {string|number} value
 * @returns {boolean}
 */
export const isBlank = (value: string | number): boolean =>
	isNil(value) || (isString(value) && !value.length);

export const getRangeSelectorsFromQuery = query => {
	const rangeEnd = get(query, 'rangeEnd', '');
	const rangeKey = get(query, 'rangeKey', RangeKeyTimeRanges.Last30Days);
	const rangeStart = get(query, 'rangeStart', '');

	return {
		rangeEnd: rangeEnd === 'null' ? null : rangeEnd,
		rangeKey,
		rangeStart: rangeStart === 'null' ? null : rangeStart
	};
};

export const getSafeRangeSelectors = (
	rangeSelectors: RangeSelectors
): SafeRangeSelectors => {
	const rangeEnd = get(rangeSelectors, 'rangeEnd', null);
	const rangeKey = get(
		rangeSelectors,
		'rangeKey',
		RangeKeyTimeRanges.Last30Days
	);
	const rangeStart = get(rangeSelectors, 'rangeStart', null);

	return {
		rangeEnd: rangeEnd || null,
		rangeKey: rangeKey === 'CUSTOM' ? null : parseInt(rangeKey),
		rangeStart: rangeStart || null
	};
};

/**
 * Normalize RangeSelectors
 * @param {RawRangeSelectors}
 * @returns {RangeSelectors}
 */
export const normalizeRangeSelectors = (
	rangeSelectors: RawRangeSelectors
): RangeSelectors => {
	const {rangeEnd, rangeKey, rangeStart} = rangeSelectors;

	if (rangeEnd && rangeStart) {
		return {
			...rangeSelectors,
			rangeKey: RangeKeyTimeRanges.CustomRange
		};
	}

	return {
		rangeEnd: '',
		rangeKey: String(rangeKey) as RangeKeyTimeRanges,
		rangeStart: ''
	};
};

/**
 * Check if the value is blank and returns value.
 * @param {string|number} value
 * @param {string|number} defaultValue
 * @returns {string|number} Returns defaultValue if value is blank.
 */
export const getSafeDisplayValue = (
	value: string | number,
	defaultValue: string | number = '-'
): string | number => (isBlank(value) ? defaultValue : value);

/**
 * Create a Blob object from data string and temporarily attach
 * an anchor element to the DOM to click on and trigger download.
 */
export const downloadDataAsFile = ({
	data,
	name,
	type
}: {
	data: string;
	name: string;
	type: string;
}) => {
	const blob = new Blob([data], {type});

	const linkUrl = URL.createObjectURL(blob);
	const link = document.createElement('a');
	link.href = linkUrl;
	link.setAttribute('download', name);

	document.body.appendChild(link);

	link.click();

	link.parentNode.removeChild(link);
	URL.revokeObjectURL(linkUrl);
};

/**
 * Remove Protocol
 * @param {string} url
 */
export const removeProtocol = url =>
	decodeURIComponent(url).replace(/^http(s)?:\/\//i, '');

/**
 * Remove numbers using regex
 * @param {string} str
 */
export const removeNumbers = str => str.replace(/\d+/g, ' ');

/**
 * Remove spacing using regex
 * @param {string} str
 */
export const removeSpacing = str => str.replace(/\s+/g, '');

/**
 * Returns the percent number passing as
 * parameter the current number and total number.
 * @param {number} number1
 * @param {number} number2
 * @returns {number}
 */
export const getPercentage = (number1, number2) => {
	const result = (number1 / number2) * 100;

	return isFinite(result) ? result : 0;
};

/**
 * Return the truncate text
 * @param {string} str
 * @param {number} length
 * @param {number} ending
 */
export const truncateText = (str, length, ending) => {
	if (length == null) length = 100;
	if (ending == null) ending = '...';

	return str.length > length
		? str.substring(0, length - ending.length) + ending
		: str;
};

/**
 * Is Ellipsis Active
 * @param {object} event
 */
export const isEllipisActive = ({target}) =>
	target.offsetWidth < target.scrollWidth;

/**
 * Get Align Position
 * @param {string} source
 * @param {string} target
 * @param {string} suggestedPosition
 */
export const getAlignPosition = (source, target, suggestedPosition) => {
	if (!suggestedPosition) {
		suggestedPosition = 'top';
	}

	const position = Align.align(
		source,
		target,
		ALIGNMENTS_MAP[suggestedPosition]
	);

	return POSITIONS[position];
};

/**
 * Trim and convert value to lowercase.
 * @param {string} value
 * @return {string} Lowercase & trimmed string.
 */
export const formatStringToLowercase: (value: string) => string = flow(
	toLower,
	trim
);

/**
 * Merges multiple refs to be used on one element
 */
export const mergeRef = (...refs) => instance =>
	refs.forEach(ref => {
		if (typeof ref === 'function') {
			ref(instance);
		} else if (ref) {
			ref.current = instance;
		}
	});
