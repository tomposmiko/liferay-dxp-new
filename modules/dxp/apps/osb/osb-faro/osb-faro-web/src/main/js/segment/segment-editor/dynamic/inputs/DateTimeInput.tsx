import autobind from 'autobind-decorator';
import DateInput from 'shared/components/DateInput';
import Form from 'shared/components/form';
import React from 'react';
import {formatDateToTimeZone} from 'shared/util/date';
import {
	INPUT_DATE_TIME_FORMAT,
	INPUT_DISPLAY_DATE_TIME_FORMAT,
	PropertyTypes
} from '../utils/constants';
import {ISegmentEditorInputBase} from '../utils/types';

interface IDateTimeInputProps extends ISegmentEditorInputBase {
	value: string;
}

export default class DateTimeInput extends React.Component<IDateTimeInputProps> {
	componentDidMount() {
		const {
			id,
			property: {entityName, type},
			valid
		} = this.props;

		if (!id && valid) {
			analytics.track('Dynamic Segment Creation - Completed Attribute', {
				entityName,
				type
			});
		}
	}

	@autobind
	handleDateChange(value) {
		this.props.onChange({
			type: PropertyTypes.Date,
			value: formatDateToTimeZone(value, null, 'UTC')
		});
	}

	render() {
		const {
			className,
			displayValue,
			operatorRenderer: OperatorDropdown,
			property: {entityName},
			timeZoneId,
			value
		} = this.props;

		const date = formatDateToTimeZone(
			value,
			INPUT_DATE_TIME_FORMAT,
			timeZoneId
		);

		return (
			<div className='criteria-statement'>
				<Form.Group autoFit>
					<Form.GroupItem className='entity-name' label shrink>
						{entityName}
					</Form.GroupItem>

					<Form.GroupItem className='display-value' label shrink>
						{displayValue}
					</Form.GroupItem>

					<OperatorDropdown />

					<Form.GroupItem>
						<DateInput
							className={className}
							displayFormat={INPUT_DISPLAY_DATE_TIME_FORMAT}
							format={INPUT_DATE_TIME_FORMAT}
							onDateInputChange={this.handleDateChange}
							showTimeSelector
							timeZoneId={timeZoneId}
							value={date}
						/>
					</Form.GroupItem>
				</Form.Group>
			</div>
		);
	}
}
