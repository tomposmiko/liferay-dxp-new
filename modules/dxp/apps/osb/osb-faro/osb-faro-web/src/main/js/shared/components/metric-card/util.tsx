import MetricValue from './MetricValue';
import React, {Fragment} from 'react';
import Trend from 'shared/components/Trend';
import {CHART_COLOR_NAMES} from 'shared/components/Chart';
import {get, last} from 'lodash';
import {
	getAxisFormatter,
	getDataFormatter,
	getIntervals,
	getMetricFormatter
} from 'shared/util/charts';
import {getIcon, getStatsColor} from 'shared/util/metrics';
import {INTERVAL_KEY_MAP} from 'shared/util/time';
import {MetricType} from './metrics';
import {toRounded} from 'shared/util/numbers';
import {toUnix} from 'shared/util/date';

const CHART_DATA_ID_1 = 'data_1';
const CHART_DATA_ID_2 = 'data_2';
const CHART_DATA_PREVIOUS = 'data_previous';
const METRIC_TOOLTIP_LABEL_MAP = {
	bounceRateMetric: Liferay.Language.get('avg-bounce')
};

const {
	martell: CHART_GREEN,
	martellL2: CHART_GREEN_L2,
	mormont: CHART_ORANGE,
	stark: CHART_BLUE,
	starkL2: CHART_BLUE_L2
} = CHART_COLOR_NAMES;

export const Icons = {
	negative: 'caret-bottom-l',
	neutral: undefined,
	positive: 'caret-top-l'
};

const PREVIOUS_PERIOD_VISITORS_COLOR = '#393A4A';

type TBuildTabs = (props: {
	activeItemIndex: number;
	items: {
		content: {
			details: {
				color: string;
				icon: string;
				label: string;
			};
			title: string;
			type: MetricType;
			value: string;
		};
	}[];
	onActiveItemIndexChange: (index: number) => void;
}) => {
	onClick: () => void | ((index: number) => void);
	secondaryInfo: React.ReactElement;
}[];

export const buildTabs: TBuildTabs = ({
	activeItemIndex,
	items,
	onActiveItemIndexChange
}) =>
	items.map(({content}, index) => {
		const {details, title, type, value} = content;
		const {color, icon, label} = details;

		return {
			onClick: () => {
				if (activeItemIndex !== index && !onActiveItemIndexChange)
					return () => {};

				return onActiveItemIndexChange(index);
			},
			secondaryInfo: (
				<span>
					<span className='primary-content'>
						<MetricValue type={type} value={value} />
					</span>

					{label && <Trend color={color} icon={icon} label={label} />}
				</span>
			),
			tabId: index,
			title
		};
	});

export const getMetricName = (activeItemIndex, metrics) =>
	metrics.map(({name}) => name)[activeItemIndex];

export const getActiveItem = (retVal, compareToPrevious) => {
	if (!retVal) {
		return {
			chartData: [],
			intervals: [],
			timeline: []
		};
	}

	const chartData = retVal.data.slice(0, -1);
	const timeline = last(retVal.data);

	if (!compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data: dataSet.data.slice(1)
			})),
			intervals: retVal.intervals.slice(1),

			// @ts-ignore

			timeline: {data: timeline.data.slice(1), id: timeline.id}
		};

		if (retVal.compositeData) {
			const compositeDataKeys = Object.keys(retVal.compositeData);

			retVal = {
				...retVal,
				compositeData: compositeDataKeys.reduce((acc, val) => {
					acc = {
						...acc,
						[val]: retVal.compositeData[val].slice(1)
					};

					return acc;
				}, {})
			};
		}
	} else if (compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data:
					dataSet.id !== 'data_previous'
						? [null, ...dataSet.data.slice(1)]
						: dataSet.data
			})),
			timeline
		};
	} else {
		retVal = {
			...retVal,
			chartData,
			timeline
		};
	}

	return retVal;
};

export const getPreviousValueFromCompositeData = (
	compositeData,
	dataName,
	dateKey
) => {
	const data = get(compositeData, dataName);

	if (data) {
		return data.find(val => toUnix(val.key) === dateKey).previousValue;
	}
};

export const getRegexType = (type: MetricType): RegExp => {
	if (type === MetricType.Ratings) {
		return /([/][0-9]+)/g;
	} else {
		return /([a-zA-Z%])+/g;
	}
};

export const formatValue = (
	value: string,
	regex: RegExp
): React.ReactElement[] => {
	const items = value.split(' ');

	return items.map((item, i) => {
		const [value, unit] = item.split(regex);

		return (
			<Fragment key={i}>
				{value}

				<span className='metric-value-letter'>{unit}</span>
			</Fragment>
		);
	});
};

export const getMetricCardTabsData = (result, metrics) =>
	metrics.map(({name, title, type}) => {
		const metricFormatter = getMetricFormatter(type);

		return {
			content: {
				details: {
					color: getStatsColor(
						result[name].trend.trendClassification
					),
					icon: getIcon(result[name].trend.percentage),
					label: `${toRounded(
						Math.abs(result[name].trend.percentage)
					)}%`
				},
				name,
				title,
				type,
				value: metricFormatter(result[name].value)
			}
		};
	});

export const getMetricsData = (
	result,
	metrics,
	rangeSelectors = {},
	chartDataMapFn = getMetricsChartData,
	interval = INTERVAL_KEY_MAP.day
) =>
	metrics.map(({compositeMetrics, name, title, tooltipTitle, type}) =>
		getMetricData({
			chartDataMapFn,
			compositeMetrics,
			interval,
			name,
			rangeSelectors,
			result,
			title,
			tooltipTitle,
			type
		})
	);

export const getMetricsChartData = ({
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) => [
	{
		color: CHART_BLUE,
		data: getDataFormatter(type)(histogram.map(({value}) => value)),
		id: CHART_DATA_ID_1,
		name: tooltipTitle || METRIC_TOOLTIP_LABEL_MAP[name] || title,
		tooltipTitle
	},
	{
		color: CHART_BLUE_L2,
		data: getDataFormatter(type)(
			histogram.map(({previousValue}) => previousValue)
		),
		id: CHART_DATA_PREVIOUS,
		name: Liferay.Language.get('previous-period')
	},
	{
		data: histogram.map(({key}) => key),
		id: 'x'
	}
];

export const convertHistogramKeysToDate = ({
	key,
	previousValueKey,
	valueKey,
	...otherParams
}) => ({
	key: toUnix(key),
	previousValueKey: previousValueKey.split('/').map(toUnix),
	valueKey: valueKey.split('/').map(toUnix),
	...otherParams
});

export const getMetricData = ({
	chartDataMapFn = getMetricsChartData,
	compositeMetrics,
	interval = INTERVAL_KEY_MAP.day,
	name,
	rangeSelectors,
	result,
	title,
	tooltipTitle,
	type
}) => {
	const metricFormatter = getMetricFormatter(type);

	const histogram = result[name].histogram.metrics.map(
		convertHistogramKeysToDate
	);

	const compositeData = compositeMetrics
		? {
				compositeContent: compositeMetrics.reduce(
					(
						acc,
						{
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType
						}
					) => {
						acc[compositeMetricName] = {
							details: {
								color: getStatsColor(
									result[compositeMetricName].trend
										.trendClassification
								),
								icon: getIcon(
									result[compositeMetricName].trend.percentage
								),
								label: `${toRounded(
									Math.abs(
										result[compositeMetricName].trend
											.percentage
									)
								)}%`
							},
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType,
							value: metricFormatter(
								result[compositeMetricName].value
							)
						};

						return acc;
					},
					{}
				),
				compositeData: compositeMetrics.reduce(
					(acc, {name: compositeMetricName}) => {
						acc[compositeMetricName] =
							result[compositeMetricName].histogram.metrics;

						return acc;
					},
					{}
				)
		  }
		: {};

	const dateKeysIMap = new Map(
		histogram.map(({key, valueKey}) => [key, valueKey])
	);

	return {
		...compositeData,
		asymmetricComparison: result[name].histogram.asymmetricComparison,
		content: {
			details: {
				color: getStatsColor(result[name].trend.trendClassification),
				icon: getIcon(result[name].trend.percentage),
				label: `${toRounded(Math.abs(result[name].trend.percentage))}%`
			},
			name,
			title,
			type,
			value: metricFormatter(result[name].value)
		},
		data: chartDataMapFn({
			...compositeData,
			histogram,
			name,
			title,
			tooltipTitle,
			type
		}),
		dateKeysIMap,
		format: getAxisFormatter(type),
		intervals: getIntervals(
			rangeSelectors.rangeKey,
			histogram.map(({key}) => key),
			interval,
			dateKeysIMap
		),
		prevDateKeysIMap: new Map(
			histogram.map(({key, previousValueKey}) => [key, previousValueKey])
		)
	};
};

export const getSiteMetricsChartData = ({
	compositeData,
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) =>
	name === 'visitorsMetric'
		? [
				{
					color: CHART_BLUE,
					data: getDataFormatter(type)(
						compositeData.knownVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'knownVisitorsMetric',
					id: CHART_DATA_ID_1,
					name: Liferay.Language.get('known-visitors'),
					tooltipTitle: Liferay.Language.get('known'),
					type: 'bar'
				},
				{
					color: CHART_ORANGE,
					data: getDataFormatter(type)(
						compositeData.anonymousVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'anonymousVisitorsMetric',
					id: CHART_DATA_ID_2,
					name: Liferay.Language.get('anonymous-visitors'),
					tooltipTitle: Liferay.Language.get('anonymous'),
					type: 'bar'
				},
				{
					color: PREVIOUS_PERIOD_VISITORS_COLOR,
					data: getDataFormatter(type)(
						histogram.map(({previousValue}) => previousValue)
					),
					id: CHART_DATA_PREVIOUS,
					name: Liferay.Language.get('previous-period'),
					type: 'line'
				},
				{
					data: histogram.map(({key}) => key),
					id: 'x'
				}
		  ]
		: getMetricsChartData({histogram, name, title, tooltipTitle, type}).map(
				data =>
					[CHART_DATA_ID_1, CHART_DATA_PREVIOUS].includes(data.id)
						? {
								...data,
								color:
									data.id === CHART_DATA_PREVIOUS
										? CHART_GREEN_L2
										: CHART_GREEN
						  }
						: data
		  );
