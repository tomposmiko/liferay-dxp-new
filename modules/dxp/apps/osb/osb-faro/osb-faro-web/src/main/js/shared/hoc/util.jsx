import ErrorDisplay from 'shared/components/ErrorDisplay';
import ErrorPage from 'shared/pages/ErrorPage';
import LoadingPage from 'shared/pages/Loading';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import React from 'react';
import Spinner from 'shared/components/Spinner';
import {compose} from 'redux';
import {get, omit} from 'lodash';
import {sub} from 'shared/util/lang';
import {toRoute} from 'shared/util/router';

/**
 * HOC for ErrorDisplay.
 * @param {Object} - Options object to pass as props to ErrorDisplay
 * @returns {Function} Returns the ErrorDisplay or Wrapped Component.
 */
export const withError =
	(options = {}) =>
	Component =>
	({error, errorProps = {}, pageDisplay = true, refetch, ...otherProps}) => {
		const otherOptions = omit(options, 'page');

		if (error) {
			return get(options, 'page', pageDisplay) ? (
				<ErrorPage {...errorProps} {...otherOptions} />
			) : (
				<ErrorDisplay
					onReload={refetch}
					spacer
					{...errorProps}
					{...otherOptions}
				/>
			);
		}

		return <Component refetch={refetch} {...otherProps} />;
	};

/**
 * HOC for NoResultsDisplay.
 * @param {Object} - Options object to pass as props to NoResultsDisplay.
 * @returns {Function} Returns the NoResultsDisplay or WrappedComponent.
 */
export const withEmpty =
	(options = {}) =>
	Component =>
	({
		data,
		error,
		loading,
		noResultsRenderer,
		noResultsRendererProps,
		total,
		...props
	}) => {
		if (((data && data.total === 0) || total === 0) && !loading && !error) {
			if (noResultsRenderer) {
				const NoResults = noResultsRenderer;

				return <NoResults />;
			}

			return (
				<NoResultsDisplay {...options} {...noResultsRendererProps} />
			);
		}

		return (
			<Component
				data={data}
				error={error}
				loading={loading}
				total={total}
				{...props}
			/>
		);
	};

/**
 * HOC for Loading display.
 * @param {Object} - Options object to pass as props to Loading component.
 * @returns {Function} Returns the Loading or WrappedComponent.
 */
export const withLoading =
	(options = {}) =>
	Component =>
	({
		alignCenter = false,
		className,
		data,
		fadeIn = true,
		inline = false,
		loading,
		pageDisplay = true,
		...otherProps
	}) => {
		if (loading) {
			return get(options, 'page', pageDisplay) ? (
				<LoadingPage
					className={className}
					fadeIn={get(options, 'fadeIn', fadeIn)}
					key='LOADING'
				/>
			) : (
				<Spinner
					alignCenter={get(options, 'alignCenter', alignCenter)}
					className={className}
					inline={get(options, 'inline', inline)}
					key='SPINNER'
					spacer={!get(options, 'inline', inline)}
				/>
			);
		}

		return <Component className={className} data={data} {...otherProps} />;
	};

export const withNull =
	(key, errorProps = {}) =>
	Component =>
	props => {
		const {entityType = Liferay.Language.get('page'), linkRoute} =
			errorProps;

		if (key && !props[key]) {
			return (
				<ErrorPage
					{...props}
					href={toRoute(linkRoute, props.router.params)}
					linkLabel={sub(Liferay.Language.get('go-to-x'), [
						entityType
					])}
					message={sub(
						Liferay.Language.get(
							'the-x-you-are-looking-for-does-not-exist'
						),
						[entityType.toLowerCase()]
					)}
					subtitle={sub(Liferay.Language.get('x-not-found'), [
						entityType
					])}
					title={Liferay.Language.get('404')}
				/>
			);
		}

		return <Component {...props} />;
	};

/**
 * HOC for displaying results.
 */
export const SafeResults = compose(
	withLoading(),
	withError()
)(({children, data}) => children(data));

export const WrapSafeResults = compose(
	withLoading(),
	withError()
)(({children}) => children);
