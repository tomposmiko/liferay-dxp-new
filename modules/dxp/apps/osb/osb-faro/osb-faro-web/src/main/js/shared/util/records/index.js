// import FaroConstants from '../constants';
import {INDIVIDUALS, PAGEVIEWS} from 'shared/util/subscriptions';
import {List, Map, Record} from 'immutable';

// const {
// 	pagination: {orderDefault}
// } = FaroConstants;

export {default as Account} from './Account';
export {default as CustomValue} from './CustomValue';
export {default as DataSource} from './DataSource';
export {default as DistributionTab} from './DistributionTab';
export {default as Individual} from './Individual';
export {default as User} from './User';
export {default as OrderParams} from './OrderParams';
export {default as Project} from './Project';
export {default as Property} from './Property';
export {default as PropertyGroup} from './PropertyGroup';
export {default as PropertySubgroup} from './PropertySubgroup';
export {default as RemoteData, remoteDataFromList} from './RemoteData';
export {default as Segment} from './Segment';
export {default as TimeZone} from './TimeZone';

export const Changeset = new Record({
	added: new Map(),
	removed: new Map()
});

export const Distribution = new Record({
	data: new List(),
	error: false,
	fieldMappingFieldName: null,
	loading: true
});

export const EntityLayout = new Record({
	contactsCardTemplatesList: new List(),
	headerContactsCardTemplates: new List(),
	id: null,
	name: '',
	type: null
});

export const Metric = new Record({
	count: 0,
	limit: 0,
	status: 0
});

// export const OrderParams = new Record({
// 	field: '',
// 	sortOrder: orderDefault
// });

export const Plan = new Record({
	addOns: new Map({
		[INDIVIDUALS]: new Map(),
		[PAGEVIEWS]: new Map()
	}),
	endDate: Date.now(),
	metrics: new Map({
		[INDIVIDUALS]: new Metric(),
		[PAGEVIEWS]: new Metric()
	}),
	name: '',
	startDate: Date.now()
});

export const SearchResults = new Record({
	disableSearch: false,
	items: new List(),
	total: 0
});
