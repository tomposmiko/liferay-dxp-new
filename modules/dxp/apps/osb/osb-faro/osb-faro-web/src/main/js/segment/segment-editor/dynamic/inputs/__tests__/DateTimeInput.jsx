import DateTimeInput from '../DateTimeInput';
import React from 'react';
import {cleanup, render} from '@testing-library/react';
import {Property} from 'shared/util/records';

jest.unmock('react-dom');

describe('DateTimeInput', () => {
	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(
			<DateTimeInput
				displayValue='Start Date Time'
				operatorRenderer={() => <div>{'operator'}</div>}
				property={new Property()}
				value='2012-12-12T00:00:00.000Z'
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
