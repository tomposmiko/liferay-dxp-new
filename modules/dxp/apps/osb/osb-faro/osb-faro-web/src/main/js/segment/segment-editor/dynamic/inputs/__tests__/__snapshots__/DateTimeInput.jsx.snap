// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeInput should render 1`] = `
<div>
  <div
    class="criteria-statement"
  >
    <div
      class="form-group form-group-autofit"
    >
      <div
        class="form-group-item entity-name form-group-item-label form-group-item-shrink"
      />
      <div
        class="form-group-item display-value form-group-item-label form-group-item-shrink"
      >
        Start Date Time
      </div>
      <div>
        operator
      </div>
      <div
        class="form-group-item"
      >
        <div
          aria-haspopup="true"
          class="input-group date-input-root"
        >
          <div
            class="input-group-item"
          >
            <input
              autocomplete="off"
              class="input-root form-control input-group-inset input-group-inset-after"
              data-testid="date-input"
              placeholder="YYYY-MM-DD HH:MM ZZ"
              type="text"
              value="2012-12-12 00:00"
            />
            <div
              class="input-group-inset-item input-group-inset-item-after"
            >
              <button
                class="button-root btn btn-unstyled"
                type="button"
              >
                <svg
                  class="icon-root lexicon-icon lexicon-icon-calendar"
                  viewBox=""
                >
                  <use
                    xlink:href="/o/osb-faro-web/dist/sprite.svg#calendar"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
